### PR TITLE
Add support for newer version of template-haskell (>= 2.18.0.0)

### LINKED ISSUE #7: Add support for newer version of template-haskell (>= 2.18.0.0)
I got this error while trying to compile the package, version 0.3.0.0.
image
After some debugging, I found out that the error was due to the fact that the type of the `Conp` data constructor from the template-haskell package have changed from being `ConP Name [Pat]` ([reference](https://hackage.haskell.org/package/template-haskell-2.17.0.0/docs/Language-Haskell-TH.html#t:Pat)) to `ConP Name [Type] [Pat]` ([reference](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/Language-Haskell-TH.html#t:Pat)) in all the versions >= 2.18.0.0.

Could you please look into this? I'm happy to make a PR to fix this, so please let me know if you'd be open to that.

Thank you.

### PR DESCRIPTION
**Pull request by quyumkehinde**
Reference: PR to fix [this issue](https://github.com/akshaymankar/polysemy-mocks/issues/7)

### CODE REVIEW

File: src/Test/Polysemy/Mock/TH.hs [hunk 1]
```diff
@@ -110,7 +110,11 @@ mkInitialReturns c =
 
 mkMockMatch :: Type -> ConLiftInfo -> Match
 mkMockMatch t c =
+#if MIN_VERSION_template_haskell(2,18,0)
+  let pat = ConP (cliConName c) (map snd (cliFunArgs c)) (map (VarP . fst) (cliFunArgs c))
```
Comment (akshaymankar) [inline src/Test/Polysemy/Mock/TH.hs]:
```suggestion
  let pat = ConP (cliConName c) [] (map (VarP . fst) (cliFunArgs c))
```

I think this expression doesn't need type application. So, maybe its better to leave this list of type applications as empty list.

File: src/Test/Polysemy/Mock/TH.hs [hunk 2]
```diff
@@ -126,7 +130,11 @@ mkMockMatch t c =
 {- ORMOLU_DISABLE -}
 mkMockToStateMatch :: Type -> ConLiftInfo -> Match
 mkMockToStateMatch t c =
+#if MIN_VERSION_template_haskell(2,18,0)
+  let pat = ConP (mockConName c) (map snd (cliFunArgs c)) (map VarP vars)
```
Comment (akshaymankar) [inline src/Test/Polysemy/Mock/TH.hs]:
```suggestion
  let pat = ConP (mockConName c) [] (map VarP vars)
```

Same as above.

File: src/Test/Polysemy/Mock/TH.hs [hunk 3]
```diff
@@ -159,7 +167,11 @@ mkMockToStateMatch t c =
 mkReturnsToStateMatch :: Type -> ConLiftInfo -> Match
 mkReturnsToStateMatch t c =
   let f = mkName "f"
+#if MIN_VERSION_template_haskell(2,18,0)
+      pat = ConP (returnsConName c) [VarT f] [VarP f]
```
Comment (akshaymankar) [inline src/Test/Polysemy/Mock/TH.hs]:
```suggestion
      pat = ConP (returnsConName c) [] [VarP f]
```

Same as above.

**Review by akshaymankar — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! Comments are inline.

File: src/Test/Polysemy/Mock/TH.hs (hunk 1 above)
Comment (quyumkehinde) [reply src/Test/Polysemy/Mock/TH.hs]: Okay, I've made the suggested changes.

**Pull request merged**